### ISSUE TITLE
Include generic value type converter

### ISSUE THREAD

**Issue by axelfontaine** [jOOQ/jOOQ#5876]
Over time we are accumulating a lot of converters that convert between String (TEXT, VARCHAR, ...) and simple immutable value types. Here is one example:

```java
public class HostedZoneIdConverter implements Converter<String, HostedZoneId> {
    @Override
    public HostedZoneId from(String databaseObject) {
        if (databaseObject == null) {
            return null;
        }
        return new HostedZoneId(databaseObject);
    }

    @Override
    public String to(HostedZoneId userObject) {
        if (userObject == null) {
            return null;
        }
        return userObject.getId();
    }

    @Override
    public Class<String> fromType() {
        return String.class;
    }

    @Override
    public Class<HostedZoneId> toType() {
        return HostedZoneId.class;
    }
}
```

The only thing changing between these converters is the "to" type.

Basically what has to be done here is pass the string to the constructor with one arg and then extract it again with a method (which could even use a standard name like "asString").

Can't this be solved out of the box in a more generic way to avoid all this boilerplate? (Using generics and reflection maybe...)

**Comment by axelfontaine** [comment]
Or auto-generate it in a way similar to the one suggested in #5877 using for example

```
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <valueTypeConverter/>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

To avoid using a new tag, special values could also be used. For example:

`<converter>VALUE_TYPE</converter>`

**Comment by lukaseder** [comment]
Indeed, there's room for improvement. As a quick win, note there's `org.jooq.impl.AbstractConverter` which helps avoid the `fromType()` and `toType()` methods.

jOOQ 3.9 added some utilities to `Converter`, e.g. [`Converter.ofNullable()`](https://www.jooq.org/javadoc/latest/org/jooq/Converter.html#ofNullable-java.lang.Class-java.lang.Class-java.util.function.Function-java.util.function.Function-), which can be used as follows:

```java
Converter<String, HostedZoneId> converter = Converter.ofNullable(
  String.class,
  HostedZoneId.class,
  t -> new HostedZoneId(t),
  u -> u.getId()
);
```

I could see two improvements here:

### 1. Add a `DelegatingConverter`:

A new type would allow for constructing a `Converter` by passing a delegating converter to it:

```java
public abstract class DelegatingConverter<T, U> extends AbstractConverter<T, U> {
  final Converter<T, U> delegate;
  DelegatingConverter(Converter<T, U> delegate) {
    super(delegate.fromType(), delegate.toType());
    this.delegate = delegate;
  }
  @Override public final U from(T t) { return delegate.from(t); }
  @Override public final T to(U u) { return delegate.to(u); }
}
```

Now, you could re-write your own converter to this:

```java
public class HostedZoneIdConverter extends DelegatingConverter<String, HostedZoneId> {
  public HostedZoneIdConverter() {
    super(Converter.ofNullable(
      String.class,
      HostedZoneId.class,
      t -> new HostedZoneId(t),
      u -> u.getId()
    ));
  }
}
```

For the time being, you can create that `DelegatingConverter` utility yourself and create a quick win here. I've registered an issue as this will be a useful feature anyway: https://github.com/jOOQ/jOOQ/issues/5878

### 2. Your suggestion

Ultimately, indeed, there's room for even more improvement. There are 3 things that need to be specified, though:

1. `<U>` (this is already specified in `<forcedType>`. `<T>` is known to the code generator)
2. A `Function<? super T, ? extends U>`
3. A `Function<? super U, ? extends T>`

I don't think it's a good idea to make the `new U(String)` constructor call a default. Sure, it works from time to time, but more often than not, there will be custom logic, just like the `u.getId()` logic is custom. So, perhaps, we could add a way to configure the content of the conversion methods? E.g.:

```xml
<!-- actual XML elements may change -->
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <nullableConverter>
      <from>t -> new HostedZoneId(t)</from>
      <to>u -> u.getId()</to>
    </nullableConverter>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

Though, would we really win anything with this?

**Comment by axelfontaine** [comment]
Actually yes we would win a lot. No we have literally dozens of those converter classes (and their associated `forcedType` xml elements) which we could collapse into just the `forcedType` elements.

Now in terms of from, the idea here is of course not the focus on the String type. In essence this type should be the database type. So we would be looking for a single-arg constructor that takes the database type and for example a method that is called "as"+the database type.

I'm fully aware this is not a general purpose solution. It is really tailored for value types. Those are however IMHO the best possible match for jOOQ as you get fantastic strong typing throughout your application by using them pervasively.

**Comment by lukaseder** [comment]
I'll need to think about this. The reason why I initially objected is because immutable, single-args-constructor classes are just one approach to value types. In particular, JDK 9 has deprecated some of these constructors, e.g. [`Boolean(String)`](http://download.java.net/java/jdk9/docs/api/java/lang/Boolean.html#Boolean-java.lang.String-) (finally, in my opinion) and favours constructing value types statically through `Boolean.valueOf(String)`.

And this still doesn't answer the concern abuot a random `to()` method implementation.

(don't get me wrong, I perfectly understand the desire to reduce this boiler plate. I just don't want to add a feature prematurely here...)

**Comment by axelfontaine** [comment]
No worries at all. I fully subscribe to the same philosophy when it comes to library design: when in doubt leave it out! (You can always add it later)

**Comment by lukaseder** [comment]
Well, if anything at all, https://github.com/jOOQ/jOOQ/issues/5878 is already added to 3.10 :)

**Comment by lukaseder** [comment]
Also useful in this context: https://github.com/jOOQ/jOOQ/issues/5884

**Comment by lukaseder** [comment]
The idea mentioned before won't be implemented as we now have #5884:

```xml
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <nullableConverter>
      <from>t -> new HostedZoneId(t)</from>
      <to>u -> u.getId()</to>
    </nullableConverter>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

Rather than implementing a new XML based mini-language, we can now put entire Java expressions as converters in the config:

```xml
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <converter>org.jooq.Converter.ofNullable(String.class, HostedZoneId.class, HostedZoneId::new, HostedZoneId::getId</converter>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

The existing `<customType/>` configuration can be used if such converter expressions need to be frequently reused.

... which leaves only the original feature request: Having a simple `<valueTypeConverter/>` flag, which does everything magically. I'm currently postponing this, as this "magic" is too opinionated on what a custom `HostedZoneId` should do (constructor instantiation, static `valueOf()`, and how to get that `getId()` method).

Perhaps in the near future, we'll come up with something more concrete.

**Comment by axelfontaine** [comment]
While both options have their merits, at the very least the `userType` should generate an import otherwise both your examples would have compile issues...

Of the two, my preference goes for the former (`nullableConverter`) for readability. In fact the `t ->` and `u ->` are always there so they could also be omitted...

**Comment by lukaseder** [comment]
> at the very least the userType should generate an import otherwise both your examples would have compile issues...

Yes it does. The example implicitly relies on this.

> Of the two, my preference goes for the former (nullableConverter) for readability. In fact the t -> and u -> are always there so they could also be omitted...

Too late :) My preference went for the more generic solution because:

1. It is *much* more powerful and extendable
2. It could immediately be used for the `EnumConverter` feature request (https://github.com/jOOQ/jOOQ/issues/5877), in case of which a `new EnumConverter(DatabaseType.class, UserType.class)` expression is generated.

So, if this approach can implement two things (and possibly more), then I'll definitely prefer that solution.

**Comment by axelfontaine** [comment]
Another idea that may be worth considering is the use of annotations similar to Jackson's @JsonCreator (for the single argument static method or constructor that does json field -> value type) and `@JsonValue` (for the zero-argument instance method that returns value type -> json field).

Example:

```
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <valueTypeConverter/>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

```
public class HostedZoneId {
    private final String id;

    @JooqCreator
    public HostedZoneId(String id) {
        this.id = id;
    }

    @JooqValue
    public String getId() {
        return id;
    }
}
```

The value type converter could then auto-discover and invoke those methods at runtime. If performance is an issue caching could always be added later.

**Comment by lukaseder** [comment]
There are two ways to implement that suggestion:

- jOOQ could actually need to class-load the `com.abc.HostedZoneId` type to introspect it. This would allow for generating the mapping code directly in the generated table. But that would be quite a strategic shift...
- There could be a mini-library for data type conversion that would implement the semantics you're suggesting. Quite an interesting project, I've thought about extracting jOOQ's auto-conversion utilities a few times (e.g. automatically converting from `String` to `Integer`, etc.). This is definitely an idea worth following.

**Comment by axelfontaine** [comment]
Actually this could even be part of the main library at first and be extracted later if desired.

**Comment by lukaseder** [comment]
Yes, that's what I meant.

**Comment by lukaseder** [comment]
Reviewing this thread now.

It is easy to overengineer things in the name of convenience. This thread contains a few suggestions where the number of questions raised is bigger than the number of questions answered (e.g. adding annotations to converters requiring the converters to be class-loaded, which is currently not necessary, which would complicate the documentation and setup of code generation configurations).

However, I'll revise my decision here: https://github.com/jOOQ/jOOQ/issues/5876#issuecomment-280652536. This does seem very useful:

```xml
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <lambdaConverter>
      <from>t -> new HostedZoneId(t)</from>
      <to>u -> u.getId()</to>
    </lambdaConverter>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

Alternatively:

```xml
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <lambdaConverter>
      <from>HostedZoneId::new</from>
      <to>HostedZoneId::getId</to>
    </lambdaConverter>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

It saves a bit of unnecessary repetition compared to what is possible today, via #5884:

```xml
<forcedType>
    <userType>com.abc.HostedZoneId</userType>
    <converter>org.jooq.Converter.ofNullable(
        String.class, 
        HostedZoneId.class, 
        HostedZoneId::new, 
        HostedZoneId::getId
    )</converter>
    <expression>.*hosted_zone</expression>
    <types>text</types>
</forcedType>
```

I've created a new feature request for this particular idea: #10532.

I think all the topics of this issue are addressed now, so I'm closing this as a duplicate of its spin-offs.

Suggestions that won't be implemented:

- Allowing for omitting `<from/>`. While the constructor call may be a reasonable default for some data types, I don't think it's better for value types to be instantiated via constructor than e.g. via `Type::valueOf` or some other static method.
- Allowing for omitting `<to/>`. The only way we can discover the `<to/>` implementation would be via class loading the `<userType/>` and then reflecting on it. That's brittle and doesn't work well as soon as the `<userType/>` has any functionality.

**Comment by lukaseder** [comment]
Also, a lot of the pain caused by repeatedly defining similar `<forcedType/>` definitions may be better addressed via the new jOOQ 3.14 embeddable type features, including e.g.

- #10482 Embeddable domains
- #10503 Making all columns (or a regex matched subset thereof) embeddable
